sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs

```diff
@@ -0,0 +1,309 @@
+        /// resolution through a small local resolver.
+        /// <param name='scope'> The scope for the particuar operation. Can be obtained from the Www-Authenticate header.
+        /// </summary>
+
+        public async Task<string> getScope(string operation, string method, string path)
+        {
+
+            if (_acrScopes.ContainsKey(method + ">" + operation))
+            {
+                return _acrScopes[method + ">" + operation];
+            }
+
+            HttpClient runtimeClient = new HttpClient();
+            HttpResponseMessage response = null;
+            string scope;
+            try
+            {
+                response = await runtimeClient.SendAsync(new HttpRequestMessage(new HttpMethod(method), operation));
+                Dictionary<string, string> data = parseHeader(response.Headers.GetValues("Www-Authenticate").FirstOrDefault());
+                scope = data.ContainsKey("scope") ? data["scope"] : hardcodedScopes(path);
+                _acrScopes[method + ">" + operation] = scope;
+            }
+            catch (Exception e)
+            {
+                throw new Exception("Could not identify appropriate Token scope: " + e.Message);
+            }
+            return scope;
+
+        }
+
+        /// <summary>
+        /// Local resolver for endpoints that will often return no scope.
+        /// <param name='operation'> Operation for which a scope is necessary
+        /// </summary>
+        private string hardcodedScopes(string operation)
+        {
+            switch (operation)
+            {
+                case "/acr/v1/_catalog":
+                case "/v2/":
+                    return "registry:catalog:*";
+                default:
+                    throw new Exception("Could not determine appropriate scope for the specified operation");
+
+            }
+        }
+
+        /// <summary>
+        /// Meant to parse out comma separated key value pairs delineated by the = sign. Accepts strings of the
+        /// format "key=value,key2=value2..." . Note this method is meant to provide limited functionality and
+        /// is not very robust.
+        ///  </summary>
+        private Dictionary<string, string> parseHeader(string header)
+        {
+
+            Dictionary<string, string> parsed = new Dictionary<string, string>();
+
+            Regex re = new Regex(@"([\w\s]+)=""[^""]+"""); //Regex is required to allow multiple scopes like push,pull
+            MatchCollection parts = re.Matches(header);
+
+            foreach (Match part in parts)
+            {
+                string[] keyValues = part.ToString().Split('=');
+                parsed.Add(keyValues[0], headerTrim(keyValues[1]));
+            }
+
+            return parsed;
+        }
+
+        /// <summary>
+        /// Removes trailing whitespace or " characters.
+        /// </summary>
+        private string headerTrim(string toTrim)
+        {
+            toTrim = toTrim.Trim();
```

Comment:
This might throw for something like `"` (single quote)

Reply:
I am not sure I understand, I don't think Trim() throws errors.

Or are you talking about the calls to substring?

Reply:
I meant the call to substring, but now that I look at it again, it won't be an issue since the `toTrim.Length` will at worst be 0 and the `Substring` method won't do anything.